### PR TITLE
Support Date32 on dateTrunc and toStartOfInterval

### PR DESCRIPTION
**Pull request by liuneng1994**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support Date32 on dateTrunc and toStartOfInterval

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 8ca4c872b8a9c97c8ff70f6604889e1c82997ad6 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68874/8ca4c872b8a9c97c8ff70f6604889e1c82997ad6/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2888b01d644c74f6cc2c0e4cfda2a0006746f2bb/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by yariks5s — CHANGES_REQUESTED** [review]
Also, docs contain stale information about `dateTrunc` and `toStartOfInterval`, can you please extend the datatypes for arguments of these functions

File: src/Functions/date_trunc.cpp [hunk 1]
```diff
@@ -75,11 +77,11 @@ class FunctionDateTrunc : public IFunction
...
-            if (!isDate(arguments[1].type) && !isDateTime(arguments[1].type) && !isDateTime64(arguments[1].type))
+            if (!isDateOrDate32(arguments[1].type) && !isDateTime(arguments[1].type) && !isDateTime64(arguments[1].type))
                 throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of 2nd argument of function {}. "
                     "Should be a date or a date with time", arguments[1].type->getName(), getName());
 
-            second_argument_is_date = isDate(arguments[1].type);
+            second_argument_is_date = isDateOrDate32(arguments[1].type);
 
             if (second_argument_is_date && ((datepart_kind == IntervalKind::Kind::Hour)
                 || (datepart_kind == IntervalKind::Kind::Minute) || (datepart_kind == IntervalKind::Kind::Second)))
```
Comment (yariks5s) [inline src/Functions/date_trunc.cpp:87]: The exception code should be like this: throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument for function {}", arguments[1].type->getName(), getName());

**Review by yariks5s — APPROVED** [review]

File: src/Functions/date_trunc.cpp (hunk 1 above)
Comment (liuneng1994) [reply src/Functions/date_trunc.cpp:87]: done

**Comment by yariks5s** [comment]
Hi, could you please merge your branch with the master? Broken checks seem to be OK now🙂

**Pull request merged**